File: cli/src/main/java/io/bitken/ss/cli/plan/Branch.java
package io.bitken.ss.cli.plan;

import io.bitken.ss.cli.HasSpec;
import io.bitken.ss.gw.GitState;
import io.bitken.ss.svc.plan.Slugs;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Model.OptionSpec;

import java.util.concurrent.Callable;

/**
 * {@code plan branch --issue ID --desc S} or {@code plan branch --plan N --desc S}
 *
 * <p>Pass {@code --issue} (e.g. {@code pb-310}) to name the branch after an
 * external issue id; the branch is {@code t/{issue}-{slug}}. Pass {@code --plan}
 * (e.g. {@code 71}) to name it after the plan number; the branch is
 * {@code t/{N}-{slug}}. Exactly one of the two must be present.
 */
public class Branch implements Callable<Integer>, HasSpec {

    private final CommandSpec spec;
    private final GitState gitState;

    public Branch(GitState gitState) {
        this.gitState = gitState;
        spec = CommandSpec.wrapWithoutInspection(this);
        spec.name("branch");
        spec.usageMessage().description("Create a task branch locally and print the push line.");
        spec.addOption(OptionSpec.builder("--issue").paramLabel("ISSUE_ID").required(false).description("Issue ID for the branch (used in the slug; defaults to the plan number)").type(String.class).build());
        spec.addOption(OptionSpec.builder("--plan").paramLabel("PLAN_NUMBER").required(false).description("Plan number (used for the slug when --issue is absent)").type(Integer.class).build());
        spec.addOption(OptionSpec.builder("--desc").paramLabel("TEXT").required(true).description("Short branch description").type(String.class).build());
    }

    @Override
    public CommandSpec getSpec() { return spec; }

    @Override
    public Integer call() {
        var pr = spec.commandLine().getParseResult();
        String desc = pr.matchedOption("desc").getValue();

        String prefix = resolvePrefix(pr);
        if (prefix == null) {
            System.out.println("ERROR: provide exactly one of --issue or --plan");
            return 1;
        }

        String branchName = Slugs.branchName(prefix, desc);
        if (gitState.branchExists(branchName)) {
            System.out.println("ERROR: branch " + branchName + " already exists");
            return 1;
        }
        if (!gitState.createBranch(branchName)) {
            System.out.println("ERROR: failed to create branch " + branchName);
            return 1;
        }
        System.out.println("Created branch: " + branchName);
        System.out.println("Run: git push -u origin " + branchName);
        return 0;
    }

    private static String resolvePrefix(picocli.CommandLine.ParseResult pr) {
        boolean hasIssue = pr.hasMatchedOption("issue");
        boolean hasPlan  = pr.hasMatchedOption("plan");
        if (hasIssue == hasPlan) return null;
        if (hasIssue) return ((String) pr.matchedOption("issue").getValue()).toLowerCase();
        return String.valueOf((Integer) pr.matchedOption("plan").getValue());
    }
}

File: exp/rust/crates/ss-cli/src/plan/branch.rs
//! `plan branch` — create a task branch locally and print the push line.
//!
//! Port of the Java `Branch`. Takes exactly one of `--issue` / `--plan`:
//! Java computes this as `hasIssue == hasPlan` → error, so both-or-neither is
//! rejected with the same message. Enforced here at dispatch rather than with
//! a clap arg group, whose wording would differ from Java's.

use ss_core::gw::GitState;
use ss_core::plan::branch_name;

pub fn run(git_state: &GitState, issue: Option<&str>, plan: Option<u32>, desc: &str) -> i32 {
    let Some(prefix) = resolve_prefix(issue, plan) else {
        println!("ERROR: provide exactly one of --issue or --plan");
        return 1;
    };

    let branch = branch_name(&prefix, desc);
    if git_state.branch_exists(&branch) {
        println!("ERROR: branch {branch} already exists");
        return 1;
    }
    if !git_state.create_branch(&branch) {
        println!("ERROR: failed to create branch {branch}");
        return 1;
    }
    println!("Created branch: {branch}");
    println!("Run: git push -u origin {branch}");
    0
}

/// The slug prefix: the lowercased issue id, or the stringified plan number.
/// `None` when both or neither were given.
fn resolve_prefix(issue: Option<&str>, plan: Option<u32>) -> Option<String> {
    match (issue, plan) {
        (Some(issue), None) => Some(issue.to_lowercase()),
        (None, Some(plan)) => Some(plan.to_string()),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exactly_one_selector_is_required() {
        assert_eq!(resolve_prefix(Some("PB-42"), None).as_deref(), Some("pb-42"), "issue is lowercased");
        assert_eq!(resolve_prefix(None, Some(7)).as_deref(), Some("7"));
        assert_eq!(resolve_prefix(None, None), None, "neither is an error");
        assert_eq!(resolve_prefix(Some("PB-42"), Some(7)), None, "both is an error too");
    }
}
